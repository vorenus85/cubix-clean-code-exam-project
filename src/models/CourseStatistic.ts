export class CourseStatistic{
    constructor(
        private courseName:string,
        private totalLectures:number,
        private lecturesCompleted:number,
        private progress:number,
        private lastAccessed:Date
        ) { }
    }