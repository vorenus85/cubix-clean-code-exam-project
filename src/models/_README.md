The "models" folder is where classes or modules representing data structures, entities, or domain objects are stored. These classes often define the structure of the data used within the application. In the context of databases, these models might map to tables or documents.
