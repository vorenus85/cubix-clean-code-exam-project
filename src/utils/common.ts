export function delay(milisec: number): Promise<void> {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, milisec)
  })
}
