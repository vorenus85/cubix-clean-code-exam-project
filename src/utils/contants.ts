export const PERSON_ERROR_MSG = {
  NAME_EMPTY: 'Name must be non-empty',
  INVALID_EMAIL: 'Invalid email address',
}

export const COURSE_ERROR_MSG = {
  COURSE_NAME_EMPTY: 'Course Name must be non-empty',
  START_DATE_MUST_DATE: 'Start Date must be Date',
  COST_IN_HUF_MUST_POSITIVE_NUMBER: 'Cost in huf must be positive number',
  LENGTH_IN_WEEK_MUST_POSITIVE_NUMBER:
    'Length in weeks must be must be positive number',
}

export const COURSE_STATISTIC_MSG = {
  COURSE_NAME_EMPTY: 'Course Name must be non-empty',
  TOTAL_LECTURE_MUST_BE_POSITIVE: 'Total lectures must be positive number',
  LECTURE_COMPLETED_MUST_BE_POSITIVE:
    'Lectures completed must be positive number',
  PROGRESS_MUST_BE_POSITIVE: 'Progress must be positive number',
  LAST_ACCESSED_MUST_DATE: 'Last accessed must be Date',
}
