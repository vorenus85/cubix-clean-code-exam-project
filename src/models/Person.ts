import { PERSON_ERROR_MSG } from '../utils/personErrorMessages'

export abstract class Person {
  constructor(
    private name: string,
    private email: string
  ) {
    this.SetName(name)
    this.SetEmail(email)
  }

  public GetName(): string {
    return this.name
  }

  public GetEmail(): string {
    return this.email
  }

  private SetName(value: string) {
    if (!value || typeof value !== 'string') {
      throw new Error(PERSON_ERROR_MSG.NAME_EMPTY)
    } else {
      this.name = value
    }
  }

  private SetEmail(email: string) {
    if (!this.IsValidEmail(email)) {
      throw new Error(PERSON_ERROR_MSG.INVALID_EMAIL)
    } else {
      this.email = email
    }
  }

  private IsValidEmail(email: string): boolean {
    // Basic email validation using a regular expression
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
    return emailRegex.test(email)
  }
}
