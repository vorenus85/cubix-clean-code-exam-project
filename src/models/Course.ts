import{ Student} from"./Student";

export class Course{ private students:Student[] =[];
    constructor( 
        private courseName: string, 
        private startDate: Date, 
        private lengthInWeeks: number, 
        private costInHuf: number) { }
    }