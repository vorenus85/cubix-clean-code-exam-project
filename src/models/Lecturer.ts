import { Course } from './Course'
import { Person } from './Person'

export class Lecturer extends Person {
  constructor(
    name: string,
    email: string,
    public assignedCourses: string[] = []
  ) {
    super(name, email)
  }
}
