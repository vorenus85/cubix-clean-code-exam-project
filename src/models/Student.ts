import { Course } from './Course'
import { Person } from './Person'

export class Student extends Person {
  constructor(
    name: string,
    email: string,
    public registeredCourses: string[] = []
  ) {
    super(name, email)
  }
}
