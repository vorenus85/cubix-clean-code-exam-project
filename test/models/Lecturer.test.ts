import { Lecturer } from '../../src/models/Lecturer'
import { PERSON_ERROR_MSG } from '../../src/utils/contants'
import { LecturerFixture } from '../fixtures'

describe('Lecturer tests', () => {
  let lecturer: Lecturer
  const name = 'Joe'
  const email = 'joe@email.com'

  describe('Happy paths', () => {
    it('should create Lecturer with valid name and email', () => {
      // Arrange
      lecturer = new Lecturer(name, email)

      // Assert
      expect(lecturer.GetName()).toBe(name)
      expect(lecturer.GetEmail()).toBe(email)
    })

    it('should Lecturer exits with all possible properties', () => {
      const actualResult = LecturerFixture

      expect(actualResult).toMatchSnapshot()
    })
  })

  describe('Error paths', () => {
    it('should throw an error for invalid name', () => {
      const errorMessage = PERSON_ERROR_MSG.NAME_EMPTY
      const errorExpected = new Error(errorMessage)

      expect(() => new Lecturer('', email)).toThrow(errorExpected)
    })

    test.each`
      invalidEmail
      ${''}
      ${'joe.email.com'}
      ${'joe.@.com'}
    `(
      'should throw an error for invalid email given email is: $invalidEmail',
      ({ invalidEmail }) => {
        const errorMessage = PERSON_ERROR_MSG.INVALID_EMAIL
        const errorExpected = new Error(errorMessage)

        expect(() => new Lecturer(name, invalidEmail)).toThrow(errorExpected)
      }
    )
  })
})
