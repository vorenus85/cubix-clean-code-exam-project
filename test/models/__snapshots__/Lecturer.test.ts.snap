// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lecturer tests Happy paths should Lecturer exits with all possible properties 1`] = `
Lecturer {
  "assignedCourses": [
    "React",
  ],
  "email": "joe@cocker.com",
  "name": "Joe Cocker",
}
`;
